### DISCUSSION TITLE
Send data with callback_data

### DISCUSSION THREAD

**Discussion by Adrian-MorenoG** [python-telegram-bot/python-telegram-bot#2815] [Q&A] [answered]
Good night, 
I have tried to send data from function1 to function2 through the callback_data of an InlineKeyboardButton but I can't get it to work and I don't know what I am doing wrong. I basically need to get the subject_id from function1 and use it in function2.
Could you please guide me to understand the error?
Thank you very much and I love the work you do!

```python
def function1(update: Update, context: CallbackContext) -> None:
    query = update.callback_query
    query.answer()
    response_API = requests.get('http://127.0.0.1:5000/subject')
    keyboard = [
        [InlineKeyboardButton(text = response_API.json()["subjects"][subject_id]["title"], callback_data = 'function2' + '$subject_id')] for subject_id in range(len(response_API.json()["subjects"]))
    ]
    keyboard.append([InlineKeyboardButton(text = '🔙 Volver Atrás', callback_data = 'comandos')])
    query.edit_message_text(
        text = "Select...: ",
        reply_markup = InlineKeyboardMarkup(keyboard)
    )

def function2(update: Update, context: CallbackContext, subject_id) -> None:
    query = update.callback_query
    query.answer()
    response_API = requests.get('http://127.0.0.1:5000/lesson/<subject_id>')
    keyboard = [
        [InlineKeyboardButton(text = response_API.json()["lesson"][i]["title"], callback_data = 'function3')] for i in range(len(response_API.json()["lesson"]))
    ]
    keyboard.append([InlineKeyboardButton(text = '🔙 Volver Atrás', callback_data = 'function1')])
    query.edit_message_text(
        text = "Select...: ",
        reply_markup = InlineKeyboardMarkup(keyboard)
    )
```

**Comment by Bibo-Joshi** [comment, marked as answer]
Hi.
`callback_data = 'function2' + '$subject_id'` gives the constant string `'function2$subject_id'`. If you want to insert the `subject_id` into the string, you should use e.g. `f'funtion2 {subject_id}` - please look up "python string formatting" on your favorite search engine
Also in `function2` you don't even access `query.data`. 
I recommend to have a look at [this example](../blob/master/examples/inlinekeyboard.py).

**Reply by Adrian-MorenoG** [reply]
I just need to pass the info that's in subject_id from the function1 to function2 so I can use it again.
I still don't get how to do it with the example.
I have in main() everything I need, or at least that's what I think.

```python
def main():
    TOKEN = "xxxxx" # os.getenv('API_KEY')
    pp = PicklePersistence(filename='tfg_bot')
    updater = Updater(TOKEN, use_context = True, persistence=pp)
    dp = updater.dispatcher

    dp.add_handler(ConversationHandler(
        entry_points = [
            CommandHandler('start', start),
            CommandHandler('comandos', callback = comandos),
            CommandHandler('clear', clear),
            CallbackQueryHandler(function1, pattern = 'function1'),
            CallbackQueryHandler(function2, pattern = 'function2'),
        ]))
    dp.add_error_handler(error_callback)
    
    updater.start_polling()
    updater.idle()
```

**Reply by Bibo-Joshi** [reply]
Please do this:

1. remove the parameter `subject_id` from `function2`
2. insert `print('the callback query handled by function2 contains the following data:', repr(query.data))` into `function2`

then run your code and see what happens. Try passing different things to the `callback_data` argument in `function1` to see how this changes the output. I'm sure that you can figure it out ;)

**Reply by Adrian-MorenoG** [reply]
Thank you so much!
I understood it much better this way and now can do some stuff to make it work.
